src/webapp/components/form/FormState.ts

```diff
@@ -81,50 +85,215 @@ export type FormFieldState =
+    return field.value;
+}
+
+// UPDATES:
+
+export function updateFormStateAndApplySideEffects(
+    currentFormState: FormState,
+    updatedField: FormFieldState
+): FormState {
+    const updatedFormState = updateFormState(currentFormState, updatedField);
+
+    return applySideEffects(updatedFormState, updatedField);
+}
+
+function applySideEffects(updatedFormState: FormState, updatedField: FormFieldState): FormState {
+    if (updatedField.notApplicableFieldId) {
+        return {
+            ...applyEffectNotApplicableFieldUpdated(updatedFormState),
+        };
+    } else {
+        return updatedFormState;
+    }
+}
+
+function applyEffectNotApplicableFieldUpdated(formState: FormState): FormState {
+    return {
+        ...formState,
+        sections: applyEffectNotApplicableFieldUpdatedInSections(formState.sections),
+    };
+}
+
+function applyEffectNotApplicableFieldUpdatedInSections(
+    sectionsState: FormSectionState[]
+): FormSectionState[] {
+    return sectionsState.map(section => {
+        if (section?.subsections) {
+            return {
+                ...applyEffectNotApplicableFieldUpdatedInSection(section),
+                subsections: applyEffectNotApplicableFieldUpdatedInSections(section?.subsections),
+            };
+        } else {
+            return applyEffectNotApplicableFieldUpdatedInSection(section);
+        }
+    });
+}
+
+function applyEffectNotApplicableFieldUpdatedInSection(
+    sectionState: FormSectionState
+): FormSectionState {
+    return {
+        ...sectionState,
+        fields: sectionState.fields.map(field => {
+            if (field.hasNotApplicable) {
+                const notApplicableField = sectionState.fields.find(
+                    f => f.notApplicableFieldId === field.id
+                );
+
+                // TODO: FIXME TypeScript error returning the corresponding fieldValue type
+                const fieldValue = (
+                    notApplicableField?.value ? getEmptyValueForField(field) : getFieldValue(field)
+                ) as any;
+
+                return {
+                    ...field,
+                    value: fieldValue,
+                    disabled: !!notApplicableField?.value,
+                };
+            }
+            return field;
+        }),
+    };
+}
+
 export function updateFormState(prevFormState: FormState, updatedField: FormFieldState): FormState {
     return {
         ...prevFormState,
         sections: updateSectionsState(prevFormState.sections, updatedField),
     };
 }
 
-export function updateSectionsState(
+export function updateFormStateWithFieldErrors(
+    formState: FormState,
+    updatedField: FormFieldState,
+    fieldValidationErrors: ValidationError[]
+): FormState {
+    return {
+        ...formState,
+        sections: updateSectionsState(formState.sections, updatedField, fieldValidationErrors),
+    };
+}
+
+function updateSectionsState(
     prevFormSectionsState: FormSectionState[],
```

Comment:
also maybe in these subsequent functions, the parameter name can just be sections instead of  prevFormSectionsState ? Would that make is easier to understand?